pkg/agent/wireguard/client_linux.go

```diff
@@ -0,0 +1,240 @@
+		}
+		return err
+	}
+	if err := netlink.LinkSetUp(link); err != nil {
+		return err
+	}
+	client.wireGuardConfig.LinkIndex = link.Attrs().Index
+	wgDev, err := client.wgClient.Device(client.wireGuardConfig.Name)
+	if err != nil {
+		return err
+	}
+	client.privateKey = wgDev.PrivateKey
+	// WireGuard private key will be persistent when agent restarts. So antrea only needs to
+	// generate a new private key if it is empty (all zero).
+	if client.privateKey == zeroKey {
+		newPkey, err := wgtypes.GeneratePrivateKey()
+		if err != nil {
+			return err
+		}
+		client.privateKey = newPkey
+	}
+	cfg := wgtypes.Config{
+		PrivateKey:   &client.privateKey,
+		ListenPort:   &client.wireGuardConfig.WireGuardPort,
+		ReplacePeers: false,
+	}
+	patch, _ := json.Marshal(map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"annotations": map[string]string{
+				types.NodeWireGuardPublicKey: client.privateKey.PublicKey().String(),
+			},
+		},
+	})
+	if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		_, err := client.k8sClient.CoreV1().Nodes().Patch(context.TODO(), client.nodeName, apitypes.MergePatchType, patch, metav1.PatchOptions{})
+		return err
+	}); err != nil {
+		return err
+	}
+
+	return client.wgClient.ConfigureDevice(client.wireGuardConfig.Name, cfg)
+}
+
+func (client *client) RemoveStalePeers(existingPeerPublickeys []string) error {
+	wgdev, err := client.wgClient.Device(client.wireGuardConfig.Name)
+	if err != nil {
+		return err
+	}
+	restoredPeerPublicKeys := make(map[wgtypes.Key]struct{})
+	for _, peer := range wgdev.Peers {
+		restoredPeerPublicKeys[peer.PublicKey] = struct{}{}
+	}
+	for _, e := range existingPeerPublickeys {
+		pubKey, err := wgtypes.ParseKey(e)
+		if err != nil {
+			klog.ErrorS(err, "Parse WireGuard public error")
+			continue
+		}
+		delete(restoredPeerPublicKeys, pubKey)
+	}
+	for k := range restoredPeerPublicKeys {
+		if err := client.deletePeerByPublicKey(k); err != nil {
+			klog.ErrorS(err, "Delete WireGuard peer error")
+		}
+	}
+	return nil
+}
+
+func (client *client) UpdatePeer(nodeName, publicKeyString string, peerNodeIP net.IP, podCIDRs []*net.IPNet) error {
+	pubKey, err := wgtypes.ParseKey(publicKeyString)
+	if err != nil {
+		return err
+	}
+	var allowedIPs []net.IPNet
+	peerNodeIPv4, peerNodeIPv6 := peerNodeIP.To4(), peerNodeIP.To16()
+
+	if peerNodeIPv4 != nil {
+		allowedIPs = append(allowedIPs, net.IPNet{
+			IP:   peerNodeIPv4,
+			Mask: net.CIDRMask(net.IPv4len*8, net.IPv4len*8),
+		})
+	} else if peerNodeIPv6 != nil {
+		allowedIPs = append(allowedIPs, net.IPNet{
+			IP:   peerNodeIPv6,
+			Mask: net.CIDRMask(net.IPv6len*8, net.IPv6len*8),
+		})
+	} else {
+		return fmt.Errorf("peer Node IP is not valid: %s", peerNodeIP.String())
+	}
+
+	for _, cidr := range podCIDRs {
+		allowedIPs = append(allowedIPs, *cidr)
+	}
+
+	if key, exist := client.peerPublicKeyByNodeName.Load(nodeName); exist {
```

Comment:
If a Node's annotation is updated after `RemoveStalePeers` and before `UpdatePeer`. It will not be removed as `peerPublicKeyByNodeName` doesn't have the Node. Will it lead to traffic issue since there are two peers having same endpointUDP and allowedIPs but different publicKey?

Reply:
I changed it a little bit to save the known node name and public key mappings in `RemoveStalePeers`.